src/Tools/ExternalAccess/Razor/RazorAnalyzerAssemblyResolver.cs

```diff
@@ -1,58 +1,50 @@
 
 using System;
-using System.Collections.Generic;
 using System.Composition;
-using System.Diagnostics;
+using System.IO;
+using System.Linq;
 using System.Reflection;
+using System.Runtime.Loader;
 using Microsoft.CodeAnalysis.Host.Mef;
-using Microsoft.CodeAnalysis.PooledObjects;
 using Microsoft.VisualStudio.Composition;
 
 namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
 {
     [Export(typeof(IAnalyzerAssemblyResolver)), Shared]
     [method: ImportingConstructor]
     [method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-    internal class RazorAnalyzerAssemblyResolver() : IAnalyzerAssemblyResolver
+    internal sealed class RazorAnalyzerAssemblyResolver() : IAnalyzerAssemblyResolver
     {
-        private static Func<AssemblyName, Assembly?>? s_assemblyResolver;
+        private const string RazorCompilerAssemblyName = "Microsoft.CodeAnalysis.Razor.Compiler";
+        private const string RazorUtilsAssemblyName = "Microsoft.AspNetCore.Razor.Utilities.Shared";
+        private const string ObjectPoolAssemblyName = "Microsoft.Extensions.ObjectPool";
 
-        private static readonly HashSet<AssemblyName> s_assembliesRequested = [];
+        private static readonly object s_loaderLock = new();
 
-        private static readonly object s_resolverLock = new();
-
-        /// <summary>
-        /// Attempts to set the assembly resolver. Will only succeed if the <paramref name="canaryAssembly"/> has not already been requested to load and the resolver has not been set previously.
-        /// </summary>
-        /// <param name="resolver">The resolver function to set.</param>
-        /// <param name="canaryAssembly">The name of an assembly that is checked to see if it has been requested to load. Setting the resolver will fail if this has already been requested.</param>
-        /// <returns><c>true</c> when the resolver was successfully set.</returns>
-        internal static bool TrySetAssemblyResolver(Func<AssemblyName, Assembly?> resolver, AssemblyName canaryAssembly)
+        public static Assembly? ResolveRazorAssembly(AssemblyName assemblyName, string rootDirectory)
         {
-            lock (s_resolverLock)
+            if (assemblyName.Name is RazorCompilerAssemblyName or RazorUtilsAssemblyName or ObjectPoolAssemblyName)
             {
-                if (s_assemblyResolver is null && !s_assembliesRequested.Contains(canaryAssembly))
+                lock (s_loaderLock)
                 {
-                    s_assemblyResolver = resolver;
-                    return true;
+                    var compilerContext = AssemblyLoadContext.GetLoadContext(typeof(Compilation).Assembly)!;
+                    if (compilerContext.Assemblies.SingleOrDefault(a => a.GetName().Name == assemblyName.Name) is Assembly loadedAssembly)
+                    {
+                        return loadedAssembly;
+                    }
+
+                    var assembly = Path.Combine(rootDirectory, $"{assemblyName.Name}.dll");
```

Comment:
Think that this code only works when the IDE has normalized the path to the razor assemblies. Essentially this logic only works if _every_ path to the razor compiler passed via `/analyzer:` is the same path once we get to this layer. If there is a different path then we will be in a "first one wins" situation (the first winner will be cached in the compiler ALC). 

If that is correct think we should add some explanation here.

Reply:
@jaredpar I filed #76868 to track addressing this feedback.